### PR TITLE
Change m_bytes to unsigned in FixedBytesType

### LINKED ISSUE #3965: Change m_bytes to unsigned in FixedBytesType
The member variable `m_bytes` is signed, although it can never be negative. We should change its type to be unsigned.

### PR DESCRIPTION
**Pull request by djl-gg**
Fixes #3965.
Closes #4036.
Closes #4010.

🙏 sorry i left a mess on that other PR.

any chance i could poke one of yall to help me set up the local testing environment?  realized changing numBytes might lead me down a bit of a rabbit hole, and i didn't want to continuously push!  i'll buy yall a round of beers

### CODE REVIEW

**Comment by axic** [comment]
You can also click on the Circleci log and find out the reason for failures:
```
/root/project/libsolidity/ast/Types.cpp: In member function 'virtual bool dev::solidity::RationalNumberType::isImplicitlyConvertibleTo(const dev::solidity::Type&) const':
/root/project/libsolidity/ast/Types.cpp:876:38: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     return fixedBytes.numBytes() * 8 >= integerType()->numBits();
            ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Please adjust the code for it to compile.

**Comment by djl-gg** [comment]
cheers @axic 

I updated the PR to fix any compilation errors.  I don't mind not receiving any bounty for this, although I'd love to contribute more slowly on the easier tickets.

**Review by axic — CHANGES_REQUESTED** [review]
Please rebase this over #4036.

File: libsolidity/ast/Types.h [hunk 1]
```diff
@@ -365,7 +365,7 @@ class FixedPointType: public Type
 	};
 	virtual Category category() const override { return Category::FixedPoint; }
 
-	explicit FixedPointType(int _totalBits, int _fractionalDigits, Modifier _modifier = Modifier::Unsigned);
+	explicit FixedPointType(unsigned _totalBits, int _fractionalDigits, Modifier _modifier = Modifier::Unsigned);
```
Comment (axic) [inline libsolidity/ast/Types.h]: Why did you chose to make `totalBits` unsigned while leaving `fractionalDigits` signed?

Comment (djl-gg) [reply libsolidity/ast/Types.h]: I just changed everything that `m_bytes` was connected to from `int` to `unsigned` along with warning/errors until the compiler was happy.  Should I update `_factionalDigits` to unsigned as well?

**Comment by chriseth** [comment]
There is another `int` in CompilerUtils.cpp:691 and CompilerUtils.cpp:799, but then I think that's almost it!

File: libsolidity/ast/Types.h (hunk 1 above)
Comment (axic) [reply libsolidity/ast/Types.h]: If you rebase it over #4036 then you don't need to.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Sorry, but there is at least one compiler error:
```
/root/project/libsolidity/ast/Types.cpp:603:5: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
   0 <= m_fractionalDigits && m_fractionalDigits <= 80,
   ~~^~~~
```

**Comment by djl-gg** [comment]
nice catch

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -600,7 +600,7 @@ FixedPointType::FixedPointType(unsigned _totalBits, unsigned _fractionalDigits,
 {
 	solAssert(
 		8 <= m_totalBits && m_totalBits <= 256 && m_totalBits % 8 == 0 &&
-		0 <= m_fractionalDigits && m_fractionalDigits <= 80,
+		0 < m_fractionalDigits && m_fractionalDigits <= 80,
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Hm, I'm not sure if we actually force `m_fractionalDigits` to be nonzero somewhere.

Comment (djl-gg) [reply libsolidity/ast/Types.cpp]: Should I keep it as an `int`?

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: No, just remove the lower bound check.

**Comment by chriseth** [comment]
Please incorporate the last commit into the appropriate bigger one.

**Review by chriseth — DISMISSED** [review]

**Comment by djl-gg** [comment]
@axic does this look good?

**Comment by chriseth** [comment]
We have a weird test failure:
```
ASSERTION FAILURE:
- file   : SolidityEndToEndTest.cpp
- line   : 3863
- message: Invalid encoded data
   Result                                                           Expectation
 X ................................................................ .........................da99.5c69723ff2b35428c8.cde458eff965dee

```
restarted the test.

**Comment by djl-gg** [comment]
thanks @chriseth !

**Review by axic — APPROVED** [review]

**Pull request merged**